rfcs/0042-config-option.md

```diff
@@ -0,0 +1,191 @@
+- Because strongswan-ctl uses options to encode its full configuration, changes like [#49197](https://github.com/NixOS/nixpkgs/pull/49197) are needed to update our options with upstream changes.
+- Pull requests like [#57036](https://github.com/NixOS/nixpkgs/pull/57036) or [#38324](https://github.com/NixOS/nixpkgs/pull/38324) are needed because users wish to have more configuration options than the ones provided.
+
+## Previous discussions
+
+- https://github.com/NixOS/nixpkgs/pull/44923#issuecomment-412393196
+- https://github.com/NixOS/nixpkgs/pull/55957#issuecomment-464561483 -> https://github.com/NixOS/nixpkgs/pull/57716
+
+## Implementations
+
+This idea has been implemented already in some places:
+- [#45470](https://github.com/NixOS/nixpkgs/pull/45470)
+- [#52096](https://github.com/NixOS/nixpkgs/pull/52096)
+- [My Murmur module](https://github.com/Infinisil/system/blob/45c3ea36651a2f4328c8a7474148f1c5ecb18e0a/config/new-modules/murmur.nix)
+
+# Detailed design
+[design]: #detailed-design
+
+For specifying configuration files to programs in NixOS options, there should be a main single option called `config` which represents the configuration of the program as a Nix value, which can then be converted to the configuration file format the program expects. This may look as follows:
+
+```nix
+{ config, lib, ... }: with lib;
+let
+
+  cfg = config.services.foo;
+  
+  # Use port specified in config or the upstream default otherwise
+  # Needed to open the correct port
+  port = cfg.config.port or 2546;
+  
+  configText = configGen.json cfg.config;
+  
+in {
+
+  options.services.foo = {
+    enable = mkEnableOption "foo service";
+    
+    config = mkOption {
+      type = lib.types.config.json;
+      default = {};
+      description = ''
+        Configuration for foo. Refer to <link xlink:href="https://example.com/docs/foo"\>
+        for documentation on the supported values.
+      '';
+    };
+  };
+  
+  config = mkIf cfg.enable {
+  
+    # Set minimal config to get service working by default
+    services.foo.config = {
+      dataPath = "/var/lib/foo";
+      logLevel = mkDefault "WARN";
+    };
+  
+    environment.etc."foo.json".text = configText;
+    
+    networking.firewall.allowedTCPPorts = [ port ];
+    # ...
+  };
+}
+```
+
+This approach solves all* of the above mentioned problems. In addition we have the following properties that work with this approach out of the box:
+- Ability to easily query arbitrary configuration values with `nix-instantiate --eval '<nixpkgs/nixos>' -A config.services.foo.config`
+- The configuration file is well formatted with the right amount of indentation everywhere
+- Usually hardcoded defaults can now be replaced by simple assignment of the `config` option, which also allows people to override those values with `mkForce`
```

Comment:
Ah yeah, I should clarify this:
- Values that are needed in order for the service to work at all on NixOS (e.g. `dataPath = /var/lib/foo`, because that's the only directory the service has access to) should be set *without* `mkDefault`, because users shouldn't be able to easily override them. `mkForce` can be used in that case, which also gives the user a sense of "You're leaving supported territory", so they probably have something to do on their own to get it to work (like making sure the directory exists and has the correct permissions in above example).
- Values that are just needed in the configuration file because the program expect any value at all (e.g. `logLevel = "DEBUG"`, because otherwise it would fail to start) should be declared *with* `mkDefault`, such that it's easy to override them.